### PR TITLE
dmd, ldc, gdc, dub: add page

### LINKED ISSUE #11314: Let's document: Dlang tools
### Command description

**Website**: https://dlang.org

- [ ] **DMD** - Oficial D Compiler (x86/amd64 only) - [cmds reference](https://dlang.org/dmd-linux.html)
- [ ] **LDC** - LLVM front-end D Compiler - [cmds reference](https://wiki.dlang.org/Using_LDC)
- [ ] **GDC** - GNU front-end D Compiler  - [cmds reference](https://wiki.dlang.org/Using_GDC)
- [ ] **DUB** - Package manager for D packages - [cmds reference](https://dub.pm/commandline)

---

## Some commands

**LDC**

Compile a source code file into an executable binary:
```bash
      ldc2 input_source.d -of=output_executable
```

Compile source code without linking:
```bash
      ldc2 -c input_source.d
```

**DUB**

Create a new Dlang project [iterative mode]:
```bash
      dub init project_name
```
Create a new D project [vibe.d template - simple http-server]:
```bash
      dub init project_name -t vibe.d
```

Create a new D project [non-iterative mode]:
```bash
      dub init project_name -n
```
Update the dependencies in D project:
```bash
      dub upgrade
```

Install dependencies specified in D project's `dub.json|dub.sdl` file:
```bash
      dub fetch
```

Build and run D project
```bash
      dub
```

<details>
<summary>dub help</summary>

```bash
dub --help
USAGE: dub [--version] [<command>] [<options...>] [-- [<application arguments...>]]

Manages the DUB project in the current directory. If the command is omitted,
DUB will default to "run". When running an application, "--" can be used to
separate DUB options from options passed to the application.

Run "dub <command> --help" to get help for a specific command.

You can use the "http_proxy" environment variable to configure a proxy server
to be used for fetching packages.


Available commands
==================

  Package creation
  ----------------
  init [<directory> [<dependency>...]]
                        Initializes an empty package skeleton

  Build, test and run
  -------------------
  run [<package>[@<version-spec>]]
                        Builds and runs a package (default command)
  build [<package>[@<version-spec>]]
                        Builds a package (uses the main package in the current
                        working directory by default)
  test [<package>[@<version-spec>]]
                        Executes the tests of the selected package
  lint [<package>[@<version-spec>]]
                        Executes the linter tests of the selected package
  generate <generator> [<package>[@<version-spec>]]
                        Generates project files using the specified generator
  describe [<package>[@<version-spec>]]
                        Prints a JSON description of the project and its
                        dependencies
  clean [<package>]     Removes intermediate build files and cached build
                        results
  dustmite <destination-path>
                        Create reduced test cases for build errors

  Package management
  ------------------
  fetch <package>[@<version-spec>]
                        Manually retrieves and caches a package
  add <package>[@<version-spec>] [<packages...>]
                        Adds dependencies to the package file.
  remove <package>[@<version-spec>]
                        Removes a cached package
  upgrade [<packages...>]
                        Forces an upgrade of the dependencies
  add-path <path>       Adds a default package search path
  remove-path <path>    Removes a package search path
  add-local <path> [<version>]
                        Adds a local package directory (e.g. a git repository)
  remove-local <path>   Removes a local package directory
  list [<package>[@<version-spec>]]
                        Prints a list of all or selected local packages dub is
                        aware of
  search <package-name> Search for available packages.
  clean-caches          Removes cached metadata
  convert               Converts the file format of the package recipe.


Common options
==============

  -h  --help            Display general or command specific help
      --root=VALUE      Path to operate in instead of the current working dir
      --registry=VALUE  Search the given registry URL first when resolving
                        dependencies. Can be specified multiple times.
                        Available registry types:
                          DUB: URL to DUB registry (default)
                          Maven: URL to Maven repository + group id containing
                          dub packages as artifacts. E.g.
                          mvn+http://localhost:8040/maven/libs-release/dubpackages
      --skip-registry=VALUE
                        Sets a mode for skipping the search on certain package
                        registry types:
                          none: Search all configured or default registries
                          (default)
                          standard: Dont search the main registry (e.g.
                          https://code.dlang.org/)
                          configured: Skip all default and user configured
                          registries
                          all: Only search registries specified with --registry
      --annotate        Do not perform any action, just print what would be
                        done
      --bare            Read only packages contained in the current directory
  -v  --verbose         Print diagnostic output
      --vverbose        Print debug output
  -q  --quiet           Only print warnings and errors
      --verror          Only print errors
      --vquiet          Print no messages
      --color=VALUE     Configure colored output. Accepted values:
                               auto: Colored output on console/terminal,
                                     unless NO_COLOR is set and non-empty
                                     (default)
                             always: Force colors enabled
                              never: Force colors disabled
      --cache=VALUE     Puts any fetched packages in the specified location
                        [local|system|user].
```
</details>

### Command details

dub can detect these three compilers, making it easier and more configurable to run the commands.

### OS

- [X] Android
- [X] Linux
- [X] macOS (osx)
- [ ] sunOS
- [X] Windows

### PR DESCRIPTION
**Pull request by Magrid0**
- [x] The page(s) are in the correct platform directories: `common`, `linux`, `osx`, `windows`, `sunos`, `android`, etc.
- [x] The page(s) have at most 8 examples.
- [x] The page description(s) have links to documentation or a homepage.
- [x] The page(s) follow the [content guidelines](/tldr-pages/tldr/blob/main/CONTRIBUTING.md#guidelines).
- [ ] The PR title conforms to the recommended [templates](/tldr-pages/tldr/blob/main/CONTRIBUTING.md#commit-message).

closes #11314

### CODE REVIEW

File: pages/linux/gdc.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# gdc
+
+> GNU front-end D Compiler.
+> More information: <https://wiki.dlang.org/Using_GDC>.
+
+- Create an executable:
+
+`gdc {{Source_Name}}.d -o {{Output_File_Name}}`
```
Comment (pixelcmtd) [inline pages/linux/gdc.md:8]: Please use the path syntax [suggested in our style guide](https://github.com/tldr-pages/tldr/blob/main/contributing-guides/style-guide.md#paths)

```suggestion
`gdc {{path/to/source.d}} -o {{path/to/output_executable}}`
```

**Review by pixelcmtd — COMMENTED** [review]
You seem to have accidentally deleted the Style Guide. Could you please `git revert` that?

About `dmd`: As it is a compiler, I think the first example, or at least an example, should be straight compiling code, like in the `gcc` page or the `gdc` page you're also suggesting.

**Comment by Magrid0** [comment]
> You seem to have accidentally deleted the Style Guide. Could you please `git revert` that?

How do I revert it from github? also it was not the original, i must've commited it accidentally while working on it

**Comment by kbdharun** [comment]
> > You seem to have accidentally deleted the Style Guide. Could you please `git revert` that?
> 
> How do I revert it from github? also it was not the original, i must've commited it accidentally while working on it

You can perform an interactive rebase. i.e `git rebase -i HEAD~15` then you drop the unnecessary commits by changing the status from `pick` to `drop` or `d` for those ones.

I will do it for you right now.

File: pages/common/dmd.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+`dmd {{path/to/source.d}}`
+
+- Generate code for all template instantiations:
+
+`dmd -allinst`
+
+- Control if bounds checking is enabled:
+
+`dmd -boundscheck=[on|safeonly|off]`
```
Comment (gutjuri) [inline pages/common/dmd.md:16]:
```suggestion
`dmd -boundscheck={{on|safeonly|off}}`
```

File: pages/common/dmd.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+- List information on all available checks:
+
+`dmd -check=[h|help|?]`
```
Comment (gutjuri) [inline pages/common/dmd.md:20]:
```suggestion
`dmd -check={{h|help|?}}`
```

File: pages/common/dub.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+- Display general or command specific help:
+
+`dub --help`
+
+- Update the dependencies in D project:
+
+`dub upgrade`
+
+- Build and run D project:
```
Comment (gutjuri) [inline pages/common/dub.md:14]:
```suggestion
- Build and run a D project:
```

File: pages/common/dub.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# dub
+
+> Package manager for D packages.
+> More information: <https://dub.pm/commandline>.
+
+- Display general or command specific help:
+
+`dub --help`
+
+- Update the dependencies in D project:
```
Comment (gutjuri) [inline pages/common/dub.md:10]:
```suggestion
- Update the dependencies in a D project:
```

File: pages/common/dub.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+`dub upgrade`
+
+- Build and run D project:
+
+`dub`
+
+- Install dependencies specified in D project's `dub.json|dub.sdl` file:
```
Comment (gutjuri) [inline pages/common/dub.md:18]:
```suggestion
- Install dependencies specified in a D project's `dub.json` or `dub.sdl` file:
```

File: pages/common/dub.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+`dub fetch`
+
+- Create a new Dlang project [iterative mode]:
```
Comment (gutjuri) [inline pages/common/dub.md:22]:
```suggestion
- Interactively create a new D project:
```

File: pages/common/dub.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+`dub init {{Project_Name}}`
```
Comment (gutjuri) [inline pages/common/dub.md:24]:
```suggestion
`dub init {{project_name}}`
```

File: pages/common/dub.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+- Create a new D project [non-iterative mode]:
+
+`dub init {{Project_Name}} -n`
```
Comment (gutjuri) [inline pages/common/dub.md:28]:
```suggestion
- Non-interactively create a new D project:

`dub init {{project_name}} -n`
```

File: pages/common/gdc.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# gdc
+
+> GNU front-end D Compiler.
```
Comment (gutjuri) [inline pages/common/gdc.md:3]:
```suggestion
> D compiler using gcc as a backend.
```

File: pages/common/ldc.md [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# ldc
+
+> LLVM front-end D compiler.
```
Comment (gutjuri) [inline pages/common/ldc.md:3]:
```suggestion
> D compiler using LLVM as a backend.
```

File: pages/common/ldc.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+> More information: <https://wiki.dlang.org/Using_LDC>.
+
+- Compile a source code file into an executable binary:
+
+`ldc2 {{Source_file}}.d -of={{Output_Executable}}`
```
Comment (gutjuri) [inline pages/common/ldc.md:8]:
```suggestion
`ldc2 {{path/to/source.d}} -of={{path/to/output_executable}}`
```

File: pages/common/ldc.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+- Compile source without linking:
+
+`ldc2 -c {{Source_file}}.d`
```
Comment (gutjuri) [inline pages/common/ldc.md:12]:
```suggestion
`ldc2 -c {{path/to/source.d}}`
```

File: pages/common/ldc.md [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+- Select the target architecture and OS:
+
+`ldc -mtriple={{Architecture-OS}} -c {{Source_file}}.d`
```
Comment (gutjuri) [inline pages/common/ldc.md:16]:
```suggestion
`ldc -mtriple={{architecture_OS}} -c {{path/to/source.d}}`
```

File: pages/common/ldc.md [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+- Get basic help instruction:
```
Comment (gutjuri) [inline pages/common/ldc.md:18]:
```suggestion
- Display help:
```

File: pages/common/ldc.md [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+`ldc2 -h`
+
+- Get complete help instruction:
```
Comment (gutjuri) [inline pages/common/ldc.md:22]:
```suggestion
- Display complete help:
```

**Review by gutjuri — CHANGES_REQUESTED** [review]
Thank you, I have some suggestions, mainly concerning parameter format and wording.

**Review by gutjuri — COMMENTED** [review]
There are still changes that I have requested. Just noting in case you have overlooked them

**Comment by kbdharun** [comment]
Note: Don't merge this PR yet, I have some changes to do to the pages.

**Comment by gutjuri** [comment]
closes #11314

**Review by kbdharun — APPROVED** [review]

**Comment by kbdharun** [comment]
For the DMD link, we can use the generic URL. As the website redirects you to the correct documentation for your host platform.

**Review by gutjuri — APPROVED** [review]

**Pull request merged**